blockchain/chain/src/store/chain_store.rs

```diff
@@ -246,33 +188,46 @@ where
         }
         Ok(())
     }
+}
 
-    /// Returns the weight of provided tipset
-    fn weight(&self, ts: &Tipset) -> Result<BigUint, String> {
-        let mut tpow = BigUint::zero();
-        let state = StateTree::new_from_root(self.db.as_ref(), ts.parent_state())?;
-        if let Some(act) = state.get_actor(&*STORAGE_POWER_ACTOR_ADDR)? {
-            if let Some(state) = self
-                .db
-                .get::<PowerState>(&act.state)
-                .map_err(|e| e.to_string())?
-            {
-                tpow = state.total_network_power;
-            }
+fn set_genesis<DB>(db: &DB, header: BlockHeader) -> Result<(), Error>
+where
+    DB: BlockStore,
+{
+    db.write(GENESIS_KEY, header.marshal_cbor()?)?;
+    Ok(persist_headers(db, &[header])?)
+}
+
+fn persist_headers<DB>(db: &DB, bh: &[BlockHeader]) -> Result<(), Error>
+where
+    DB: BlockStore,
+{
+    let mut raw_header_data = Vec::new();
+    let mut keys = Vec::new();
+    // loop through block to push blockheader raw data and cid into vector to be stored
+    for header in bh {
+        if !db.exists(header.cid().key())? {
+            raw_header_data.push(header.marshal_cbor()?);
+            keys.push(header.cid().key());
         }
-        let log2_p = if tpow > BigUint::zero() {
-            BigUint::from(tpow.bits() - 1)
-        } else {
-            return Err("All power in the net is gone. You network might be disconnected, or the net is dead!".to_owned());
-        };
+    }
+
+    Ok(db.bulk_write(&keys, &raw_header_data)?)
+}
 
-        let mut out = ts.weight() + (&log2_p << 8);
-        let e_weight =
-            ((log2_p * BigUint::from(ts.blocks().len())) * BigUint::from(W_RATIO_NUM)) << 8;
-        let value = e_weight / (BigUint::from(BLOCKS_PER_EPOCH) * BigUint::from(W_RATIO_DEN));
-        out += &value;
-        Ok(out)
+fn put_messages<DB, T: Cbor>(db: &DB, msgs: &[T]) -> Result<(), Error>
+where
+    DB: BlockStore,
+{
+    for m in msgs {
+        let key = m.cid()?.key();
+        let value = &m.marshal_cbor()?;
+        if db.exists(&key)? {
+            return Ok(());
+        }
+        db.write(&key, value)?
```

Comment:
Idk how this was missed on initial review, if you could pls fix :)

```suggestion
        if !db.exists(&key)? {
            db.write(&key, value)?
        }
```